library/std/src/sys/unix/futex.rs

```diff
@@ -25,6 +32,29 @@ pub fn futex_wait(futex: &AtomicI32, expected: i32, timeout: Option<Duration>) {
     }
 }
 
+#[cfg(target_os = "emscripten")]
+pub fn futex_wait(futex: &AtomicI32, expected: i32, timeout: Option<Duration>) {
+    extern "C" {
+        fn emscripten_futex_wait(
+            addr: *const AtomicI32,
+            val: libc::c_uint,
+            max_wait_ms: libc::c_double,
+        ) -> libc::c_int;
+    }
+
+    let timeout_ms = timeout.map(|d| d.as_millis());
```

Comment:
as `emscripten_futex_wait` is taking a double i think we should better use `as_secs_f64` here?

```suggestion
    let timeout_ms = timeout.map_or(INFINITY, |d| d.as_secs_f64() * 1e-3);
```

although [Atomics.wait](https://tc39.es/ecma262/#sec-atomics.wait) → [SuspendAgent](https://tc39.es/ecma262/#sec-suspendagent) suggests that the *timeout* is always a non-negative integer (but at the same time accepting +&infin;<sub>𝔽</sub>), it is not clear that the actual implementation *can* accept fractional milliseconds. For instance [V8](https://github.com/v8/v8/blob/f1ae68351f19069346e623badf42be0e4385633b/src/execution/futex-emulation.cc#L321-L324) seems to accept waiting for 0.5ms as it just converts to nanoseconds without rounding.

Reply:
Great catch, indeed this was turning the timeout into a nonsensical `u64` value instead of the needed `f64`. 😅 

As for fractional milliseconds, I'd say it's best to leave that up to Emscripten / the VM impls to handle.